### PR TITLE
WebRequest deprecation

### PR DESCRIPTION
**Pull request by aik-jahoda**
## Summary

Deprecation of WebRequest

[Preview link](https://review.docs.microsoft.com/en-us/dotnet/core/compatibility/networking/6.0/webrequest-deprecated?branch=pr-en-us-20310)

Part of dotnet/runtime#33125

### CODE REVIEW

File: docs/core/compatibility/fx-core.md [hunk 1]
```diff
@@ -67,10 +67,14 @@ If you're migrating an app from .NET Framework to .NET Core, the breaking change
 
 - [WebClient.CancelAsync doesn't always cancel immediately](#webclientcancelasync-doesnt-always-cancel-immediately)
 
+- [WebRequest in not recomended for new development](#webRequest-in-not-recomended-for-new-development)
+
 ### .NET Core 2.0
 
 [!INCLUDE [behavior-change-webclient-cancelasync](../../../includes/core-changes/networking/2.0/behavior-change-webclient-cancelasync.md)]
```
Comment (aik-jahoda) [inline docs/core/compatibility/fx-core.md:74]: Consider merge `behavior-change-webclient-cancelasync.md` chapter with `behavior-change-webrequest-deprecate.md`

**Comment by aik-jahoda** [comment]
@gewarren @karelz, this is potentially sensitive topic and need explain reasons why we decided to deprecate WebRequest api. Can you please review this draft and add suggestions how to send correct message to customers?

File: docs/core/compatibility/3.1-5.0.md [hunk 2]
```diff
@@ -236,6 +236,7 @@ If you're migrating from version 3.1 of .NET Core, ASP.NET Core, or EF Core to v
 
 - [No A/W suffix probing on non-Windows platforms](#no-aw-suffix-probing-on-non-windows-platforms)
 - [Built-in support for WinRT is removed from .NET](#built-in-support-for-winrt-is-removed-from-net)
+- [WebRequest, WebClient and ServicePoint are deprecated](#webrequest_webclient_and_servicepoint_are_deprecated)
```
Comment (Youssef1313) [inline docs/core/compatibility/3.1-5.0.md]:
```suggestion
- [WebRequest, WebClient, and ServicePoint are deprecated](#webrequest-webclient-and-servicepoint-are-deprecated)
```

File: docs/core/compatibility/3.1-5.0.md [hunk 3]
```diff
@@ -258,6 +259,9 @@ If you're migrating from version 3.1 of .NET Core, ASP.NET Core, or EF Core to v
 
 ***
 
+[!INCLUDE [webrequest_webclient_and_servicepoint_are_deprecated](../../../includes/core-changes/networking/5.0/webrequest-deprecate.md)]
+
+
```
Comment (Youssef1313) [inline docs/core/compatibility/3.1-5.0.md:263]:
```suggestion
***
```

File: includes/core-changes/networking/2.0/behavior-change-webrequest-deprecate.md [hunk 4]
```diff
@@ -0,0 +1,34 @@
+### WebRequest is not recomended for new development
+
+Starting in .NET Core 2.0, <xref:System.Net.WebRequest>, <xref:System.Net.WebClient> and <xref:System.Net.ServicePoint> support was added to .NET Core in version 2.0 for backward compatibility. The reason for returning the API is for API backquard compatibility, however there are considerable runtime breaking changes.
```
Comment (Youssef1313) [inline includes/core-changes/networking/2.0/behavior-change-webrequest-deprecate.md:3]:
```suggestion
Starting in .NET Core 2.0, <xref:System.Net.WebRequest>, <xref:System.Net.WebClient>, and <xref:System.Net.ServicePoint> support was added to .NET Core in version 2.0 for backward compatibility. The reason for returning the API is for API backward compatibility, however there are considerable runtime breaking changes.
```

File: includes/core-changes/networking/2.0/behavior-change-webrequest-deprecate.md [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+#### Change description
+
+For example:
+
+ 1. WebRequest.GetRequestStream alocate memory for whole response
```
Comment (Youssef1313) [inline includes/core-changes/networking/2.0/behavior-change-webrequest-deprecate.md:9]:
```suggestion
1. WebRequest.GetRequestStream allocate memory for whole response
```

File: includes/core-changes/networking/2.0/behavior-change-webrequest-deprecate.md [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+ 2. WebClient.CancelAsync doesn't always cancel immediately
```
Comment (Youssef1313) [inline includes/core-changes/networking/2.0/behavior-change-webrequest-deprecate.md:10]:
```suggestion
2. WebClient.CancelAsync doesn't always cancel immediately
```

File: includes/core-changes/networking/2.0/behavior-change-webrequest-deprecate.md [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+#### Version introduced
+
+2.0
+
+#### Recommended action
+
+Use the <xref:System.Net.Http.HttpClient?displayProperty=fullName> class instead of <xref:System.Net.WebClient?displayProperty=fullName>, which is deprecated.
```
Comment (Youssef1313) [inline includes/core-changes/networking/2.0/behavior-change-webrequest-deprecate.md:18]: Is WebClient deprecated in .NET Core 2.0 or .NET 5.0?

This file mentions that WebClient is deprecated in .NET Core 2.0, while the other file in the PR mentions that WebRequest, WebClient, and ServicePoint are all deprecated in .NET 5.0.

That confuses me a little bit.

File: includes/core-changes/networking/5.0/webrequest-deprecate.md [hunk 8]
```diff
@@ -0,0 +1,39 @@
+### WebRequest, WebClient and ServicePoint are deprecated
```
Comment (Youssef1313) [inline includes/core-changes/networking/5.0/webrequest-deprecate.md:1]:
```suggestion
### WebRequest, WebClient, and ServicePoint are deprecated
```

File: includes/core-changes/networking/5.0/webrequest-deprecate.md [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+
+<xref:System.Net.WebRequest>, <xref:System.Net.WebClient> and <xref:System.Net.ServicePoint> classes was marked as deprecate.
```
Comment (Youssef1313) [inline includes/core-changes/networking/5.0/webrequest-deprecate.md:3]:
```suggestion
<xref:System.Net.WebRequest>, <xref:System.Net.WebClient>, and <xref:System.Net.ServicePoint> classes was marked as deprecated.
```

File: includes/core-changes/networking/5.0/webrequest-deprecate.md [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+
+#### Version introduced
+
+5.0
+
+#### Change description
+
+<xref:System.Net.WebRequest>, <xref:System.Net.WebClient> and <xref:System.Net.ServicePoint> classes was marked as deprecate. The classes are still available but they are not recomentded for new development. Only construction methods were marked with ObsoleteAttibute to reduce number of analyser warnings.
```
Comment (Youssef1313) [inline includes/core-changes/networking/5.0/webrequest-deprecate.md:11]:
```suggestion
<xref:System.Net.WebRequest>, <xref:System.Net.WebClient>, and <xref:System.Net.ServicePoint> classes was marked as deprecated. The classes are still available but they are not recommended for new development. Only construction methods were marked with ObsoleteAttribute to reduce number of analyzer warnings.
```

File: includes/core-changes/networking/2.0/behavior-change-webrequest-deprecate.md (hunk 7 above)
Comment (aik-jahoda) [reply includes/core-changes/networking/2.0/behavior-change-webrequest-deprecate.md:18]: There are two messages:
1) We don't recommend use WebRequest api (.net core 2.0 when the classes were added)
2) Add Obsolete attribute to WebRequest api (.net 5.0)

Comment (Youssef1313) [reply includes/core-changes/networking/2.0/behavior-change-webrequest-deprecate.md:18]: I honestly see the breaking change in 5.0 only.

How "not recommending an API without obsoletion" is considered a breaking change?

Am I missing something?

File: includes/core-changes/networking/5.0/webrequest-deprecate.md [hunk 11]
```diff
@@ -0,0 +1,39 @@
+### WebRequest, WebClient, and ServicePoint are deprecated
+
+<xref:System.Net.WebRequest>, <xref:System.Net.WebClient>, and <xref:System.Net.ServicePoint> classes was marked as deprecated.
```
Comment (gewarren) [inline includes/core-changes/networking/5.0/webrequest-deprecate.md:3]:
```suggestion
<xref:System.Net.WebRequest>, <xref:System.Net.WebClient>, and <xref:System.Net.ServicePoint> classes are deprecated.
```

File: includes/core-changes/networking/5.0/webrequest-deprecate.md [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+
+#### Version introduced
+
+5.0
+
+#### Change description
+
+<xref:System.Net.WebRequest>, <xref:System.Net.WebClient>, and <xref:System.Net.ServicePoint> classes was marked as deprecated. The classes are still available but they are not recommended for new development. Only construction methods were marked with ObsoleteAttribute to reduce number of analyzer warnings.
```
Comment (gewarren) [inline includes/core-changes/networking/5.0/webrequest-deprecate.md:11]:
```suggestion
The <xref:System.Net.WebRequest>, <xref:System.Net.WebClient>, and <xref:System.Net.ServicePoint> classes were added to .NET Core in version 2.0 for backward compatibility. However, they introduced several runtime breaking changes, for example, `WebRequest.GetRequestStream` allocates memory for the whole response, and `WebClient.CancelAsync` doesn't always cancel immediately.

Starting in .NET 5.0, the <xref:System.Net.WebRequest>, <xref:System.Net.WebClient>, and <xref:System.Net.ServicePoint> classes are deprecated. The classes are still available, but they are not recommended for new development. To reduce the number of analyzer warnings, only construction methods are decorated with the <xref:System.ObsoleteAttribute> attribute.
```

File: includes/core-changes/networking/2.0/behavior-change-webrequest-deprecate.md (hunk 7 above)
Comment (gewarren) [reply includes/core-changes/networking/2.0/behavior-change-webrequest-deprecate.md:18]: I agree with Youssef here. I think we should remove this file and the link where it's included.

**Comment by gewarren** [comment]
Please also add the include link to the networking category file: https://github.com/dotnet/docs/blob/master/docs/core/compatibility/networking.md

**Comment by IEvangelist** [comment]
Hi @aik-jahoda - is this still relevant or needed? This is still marked as a draft.

**Comment by aik-jahoda** [comment]
@IEvangelist yep, it is still valid, I have retargeted it to .net 6.

@karelz - this topic is important for people moving to .net 6, can you please take a look? The classes are deprecated .net 6 so we should push this forward.

File: docs/core/compatibility/networking/6.0/webrequest-deprecated.md [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+
+- `T:System.Net.WebRequest`
+- `T:System.Net.HttpWebRequest`
+- `T:System.Net.FtpWebRequest`
+- `T:System.Net.WebClient`
+- `T:System.Net.ServicePoint`
+
+### Category
+
+Networking
```
Comment (Youssef1313) [inline docs/core/compatibility/networking/6.0/webrequest-deprecated.md:44]: Core libraries feels like a valid category too, but how the docs are organized doesn't allow multiple categories. Are there any criteria to prefer one over the other?

Comment (gewarren) [reply docs/core/compatibility/networking/6.0/webrequest-deprecated.md:44]: I think "Core .NET libraries" is for anything that doesn't fit in one of the other categories. So Networking is fine here.

**Review by gewarren — APPROVED** [review]

**Pull request merged**